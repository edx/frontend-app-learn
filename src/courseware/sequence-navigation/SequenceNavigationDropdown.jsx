import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import UnitButton from './UnitButton';

export default function SequenceNavigationDropdown({
  activeUnitId,
  onNavigate,
  showCompletion,
  unitIds,
}) {
  return (
    <Dropdown className="sequence-navigation-dropdown">
      <Dropdown.Button className="dropdown-button font-weight-normal w-100 border-right-0">
        <FormattedMessage
          defaultMessage="{current} of {total}"
          description="The title of the mobile menu for sequence navigation of units"
          id="learn.course.sequence.navigation.mobile.menu"
          values={{
            current: unitIds.indexOf(activeUnitId) + 1,
            total: unitIds.length,
          }}
        />
      </Dropdown.Button>
      <Dropdown.Menu className="w-100">
        {unitIds.map(unitId => (
          <UnitButton
            className="w-100"
            isActive={activeUnitId === unitId}
            key={unitId}
            onClick={onNavigate}
            showCompletion={showCompletion}
            showTitle
            unitId={unitId}
          />
        ))}
      </Dropdown.Menu>
    </Dropdown>
  );
}

SequenceNavigationDropdown.propTypes = {
  activeUnitId: PropTypes.string.isRequired,
  onNavigate: PropTypes.func.isRequired,
  showCompletion: PropTypes.bool.isRequired,
  unitIds: PropTypes.arrayOf(PropTypes.string).isRequired,
};
