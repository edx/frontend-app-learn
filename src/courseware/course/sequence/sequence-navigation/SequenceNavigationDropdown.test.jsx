import React from 'react';
import { fireEvent } from '@testing-library/dom';
import SequenceNavigationDropdown from './SequenceNavigationDropdown';
import { render, screen } from '../../../../test/test-utils';

describe('Sequence Navigation Dropdown', () => {
  const testUnits = ['1', '2', '3'];

  const initialState = {
    models: {
      units: testUnits.reduce(
        (acc, unitId) => Object.assign(acc, {
          [unitId]: {
            contentType: 'other',
            title: unitId,
          },
        }),
        {},
      ),
    },
  };

  const mockData = {
    unitId: '1',
    onNavigate: () => {},
    showCompletion: false,
    unitIds: testUnits,
  };

  it('renders correctly without units', () => {
    const { asFragment } = render(<SequenceNavigationDropdown
      {...mockData}
      unitIds={[]}
    />);

    expect(asFragment()).toMatchSnapshot();
  });

  testUnits.forEach(unitId => {
    it(`displays proper text for unit ${unitId} on mobile`, () => {
      render(<SequenceNavigationDropdown
        {...mockData}
        unitId={unitId}
      />, { initialState });

      expect(screen.getByRole('button')).toHaveTextContent(`${unitId} of ${testUnits.length}`);
    });
  });

  testUnits.forEach(unitId => {
    it(`marks unit ${unitId} as active`, () => {
      render(<SequenceNavigationDropdown
        {...mockData}
        unitId={unitId}
      />, { initialState });

      // Only the current unit should be marked as active.
      screen.getAllByText(/^\d$/).forEach(element => {
        if (element.textContent === unitId) {
          expect(element.parentElement).toHaveClass('active');
        } else {
          expect(element.parentElement).not.toHaveClass('active');
        }
      });
    });
  });

  it('handles the clicks', () => {
    const onNavigate = jest.fn();

    render(<SequenceNavigationDropdown
      {...mockData}
      onNavigate={onNavigate}
    />, { initialState });

    screen.getAllByText(/^\d$/).forEach(element => fireEvent.click(element));
    expect(onNavigate).toHaveBeenCalledTimes(testUnits.length);
  });
});
