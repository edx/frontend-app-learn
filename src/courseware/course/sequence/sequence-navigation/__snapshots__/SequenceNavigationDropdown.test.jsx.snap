// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence Navigation Dropdown renders correctly without units 1`] = `
<DocumentFragment>
  <div
    class="dropdown sequence-navigation-dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="dropdown-toggle btn dropdown-button font-weight-normal w-100 border-right-0"
      id="pgn__dropdown-trigger-0"
    >
      <span>
        0 of 0
      </span>
    </button>
    <div
      aria-hidden="true"
      aria-labelledby="pgn__dropdown-trigger-0"
      class="dropdown-menu w-100"
      role="menu"
    />
  </div>
</DocumentFragment>
`;
